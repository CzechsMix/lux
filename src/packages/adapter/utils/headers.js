// @flow
import entries from '../../../utils/entries';

type ChangeType = 'SET' | 'DELETE';
type ChangeData = [string, void | string];
type HandlerChange = (type: ChangeType, data: ChangeData) => void;

class Headers extends Map<string, string> {
  constructor(value: { [key: string]: string }) {
    super(entries(value));
  }

  get(key: string): void | string {
    return super.get(String(key).toLowerCase());
  }

  has(key: string): boolean {
    return super.has(String(key).toLowerCase());
  }

  set(key: string, value: string): this {
    super.set(String(key).toLowerCase(), value);
    return this;
  }

  delete(key: string): boolean {
    return super.delete(String(key).toLowerCase());
  }
}

export class RequestHeaders extends Headers {
  // eslint-disable-next-line no-unused-vars
  set(key: string, value: string): this {
    return this;
  }

  // eslint-disable-next-line no-unused-vars
  delete(key: string): boolean {
    return false;
  }
}

export class ResponseHeaders extends Headers {
  handleChange: HandlerChange;

  constructor(handleChange: HandlerChange) {
    super({});
    this.handleChange = handleChange;
  }

  set(key: string, value: string): this {
    this.handleChange('SET', [key, value]);
    return super.set(key, value);
  }

  delete(key: string): boolean {
    this.handleChange('DELETE', [key]);
    return super.delete(key);
  }
}

Object.defineProperty(ResponseHeaders.prototype, 'handleChange', {
  writable: true,
  enumerable: false,
  configurable: false,
});
