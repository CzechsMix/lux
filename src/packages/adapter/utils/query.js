// @flow
import entries from '../../../utils/entries';
import isObject from '../../../utils/is-object';

const INT = /^\d+$/;
const CSV = /^(?:[\w\d-]+)(?:,[\w\d-]+)*$/;
const NULL = /^null$/;
const BOOL = /^(?:true|false)$/;
const DATE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}(Z|\+\d{4})$/;
const TRUE = /^true$/;
const DELIMITER = /[_-\s]+/;

type Params = {
  [key: string]: any;
};

const camelize = (source: string) => (
  source
    .split(DELIMITER)
    .reduce((result, part, idx) => {
      if (part[0]) {
        const [first] = part;

        return (
          result
          + (idx === 0 ? first.toLowerCase() : first.toUpperCase())
          + part.slice(1).toLowerCase()
        );
      }

      return result;
    }, '')
);

const fromString = (source: string): any => {
  if (INT.test(source)) {
    return Number.parseInt(source, 10);
  } else if (BOOL.test(source)) {
    return TRUE.test(source);
  } else if (NULL.test(source)) {
    return null;
  } else if (CSV.test(source)) {
    return source.split(',');
  } else if (DATE.test(source)) {
    return new Date(source);
  }

  return source;
};

export const fromObject = (source: Object): Params => (
  entries(source).reduce((obj, [k, v]) => {
    const result = obj;
    const key = camelize(k);
    let value = v;

    if (typeof value === 'string') {
      value = fromString(value);
    } else if (isObject(value)) {
      value = fromObject(value);
    }

    if (key === 'include' && Array.isArray(value)) {
      value = value.map(camelize);
    } else if (key === 'fields' && isObject(value)) {
      value = entries(value).reduce((fields, [resource, names]) => (
        Object.assign(fields, {
          [resource]: Array.isArray(names) ? names.map(camelize) : names,
        })
      ), {});
    }

    result[key] = value;

    return result;
  }, {})
);
