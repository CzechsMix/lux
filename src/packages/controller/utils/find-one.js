// @flow
import merge from '../../../utils/merge';
// eslint-disable-next-line no-unused-vars
import type { Model, Query } from '../../database';
import type { Request } from '../../request';

import paramsToQuery from './params-to-query';

/**
 * @private
 */
export default function findOne<T: Model>(
  model: Class<T>,
  req: Request
): Query<T> {
  const params = merge(req.defaultParams, req.params);
  const { id, select, include } = paramsToQuery(model, params);

  return model.find(id)
    .select(...select)
    .include(include);
}
