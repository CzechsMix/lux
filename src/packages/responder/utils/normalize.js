// @flow
import { STATUS_CODES } from '../../../constants';
import stringify from '../../../utils/stringify';

import dataFor from './data-for';

type ResponseData = {
  data: string;
  statusCode: number;
};

/**
 * @private
 */
export default function normalize(content: any): ResponseData {
  let data;
  let statusCode;

  switch (typeof content) {
    case 'boolean':
      if (content) {
        statusCode = 204;
      } else {
        statusCode = 401;
        data = dataFor(statusCode);
      }
      break;

    case 'number':
      if (STATUS_CODES.has(content)) {
        statusCode = content;
      } else {
        statusCode = 404;
      }
      data = dataFor(statusCode);
      break;

    case 'object':
      if (content instanceof Error) {
        statusCode = Number.parseInt(content.statusCode, 10) || 500;
        data = dataFor(statusCode, content);
      } else if (!content) {
        statusCode = 404;
        data = dataFor(statusCode);
      } else {
        statusCode = 200;
        data = content;
      }
      break;

    case 'undefined':
      statusCode = 404;
      data = dataFor(statusCode);
      break;

    default:
      statusCode = 200;
      data = content;
  }

  data = stringify(data);

  return {
    data,
    statusCode,
  };
}
