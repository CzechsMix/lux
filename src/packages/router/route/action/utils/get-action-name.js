// @flow
import type { Request } from '../../../../request';

/**
 * @private
 */
export default function getActionName({
  route: {
    action
  }
}: Request): string {
  return action;
}
