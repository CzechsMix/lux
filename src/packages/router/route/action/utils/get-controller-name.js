// @flow
import type { Request } from '../../../../request';

/**
 * @private
 */
export default function getControllerName({
  route: {
    controller: {
      constructor: {
        name
      }
    }
  }
}: Request): string {
  return name;
}
