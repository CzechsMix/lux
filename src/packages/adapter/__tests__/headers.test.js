/* @flow */

import { MIME_TYPE } from '../../jsonapi';
import { Headers, ResponseHeaders } from '../utils/headers';

describe('module "adapters/headers"', () => {
  let subject;

  describe('util Headers', () => {
    beforeEach(() => {
      subject = new Headers();
    });

    describe('#constructor()', () => {
      beforeEach(() => {
        subject = new Headers({
          Accept: MIME_TYPE,
          'Content-Type': MIME_TYPE,
        });
      });

      it('creates an instance of Headers', () => {
        expect(subject).toMatchSnapshot();
      });
    });

    ['get', 'has'].forEach(method => {
      describe(`#${method}()`, () => {
        const key = 'Accept';
        const value = method === 'get' ? MIME_TYPE : true;

        beforeEach(() => {
          subject.set(key, MIME_TYPE);
        });

        it('returns the correct value for the given key', () => {
          // $FlowIgnore
          expect(subject[method](key)).toBe(value);
        });

        it('is not case sensitive', () => {
          // $FlowIgnore
          expect(subject[method](key.toUpperCase())).toBe(value);
        });
      });
    });

    describe('#set()', () => {
      const key = 'Accept';
      const value = MIME_TYPE;

      it('sets the correct value for the given key', () => {
        expect(subject.set(key, value)).toMatchSnapshot();
      });
    });

    describe('#delete()', () => {
      const key = 'Accept';
      const value = MIME_TYPE;

      beforeEach(() => {
        subject.set(key, value);
      });

      it('deletes the entry for a given key', () => {
        subject.delete(key);
        expect(subject.size).toBe(0);
      });

      it('is not case sensitive', () => {
        subject.delete(key.toUpperCase());
        expect(subject.size).toBe(0);
      });
    });
  });

  describe('util ResponseHeaders', () => {
    const key = 'Content-Type';
    const value = MIME_TYPE;
    const handleChange = jest.fn();

    beforeEach(() => {
      subject = new ResponseHeaders(handleChange);
    });

    describe('#set()', () => {
      it('calls the change handler', () => {
        subject.set(key, value);
        expect(handleChange).toBeCalledWith('SET', [key, value]);
      });
    });

    describe('#delete()', () => {
      it('calls the change handler', () => {
        subject.delete(key);
        expect(handleChange).toBeCalledWith('DELETE', [key, null]);
      });
    });
  });
});
