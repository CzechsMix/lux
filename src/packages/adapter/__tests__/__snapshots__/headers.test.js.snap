exports[`module "adapters/headers" util Headers #constructor() creates an instance of Headers 1`] = `
Array [
  Array [
    "accept",
    "application/vnd.api+json",
  ],
  Array [
    "content-type",
    "application/vnd.api+json",
  ],
]
`;

exports[`module "adapters/headers" util Headers #set() sets the correct value for the given key 1`] = `
Array [
  Array [
    "accept",
    "application/vnd.api+json",
  ],
]
`;
